grpc_unary_attestation/src/server.rs

```diff
@@ -18,20 +18,89 @@
 //! protocol.
 
 use crate::proto::{unary_session_server::UnarySession, UnaryRequest, UnaryResponse};
+use lru::LruCache;
+use oak_remote_attestation::handshaker::{AttestationBehavior, Encryptor, ServerHandshaker};
 use oak_utils::LogError;
+use std::sync::{Arc, Mutex};
 use tonic;
 
-/// gRPC Attestation Service implementation.
-pub struct AttestationServer<F, L: LogError> {
+type SessionId = u64;
+enum SessionState {
+    // Boxed due to large size difference, ref: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant
+    HandshakeInProgress(Box<ServerHandshaker>),
+    EncryptedMessageExchange(Encryptor),
+}
```

Comment:
This arguably duplicates the `ServerHandshakerState` state key inside of the `ServerHandshaker`, which itself is an enum with a `ServerHandshaker::Completed(AeadEncryptor)` value which corresponds `EncryptedMessageExchange(Encryptor)`.

Without additional refactoring however, `ServerHandshaker` cannot be used directly. It's `get_encryptor` method consumes the wrapping `ServerHandshaker` instance. 

Effectively this comes up again at L77 in `SessionsTracker.pop_session_state`, which checks for handshake completion when retrieving session state, as a completed `ServerHandshaker` and `Encryptor` can functionally represent the same state. (L77 checks for this, always returns the underlying `Encryptor` to ensure consistency).

Reply:
Imo the ideal architecture would be to merge `ServerHandshaker` & `Encryptor` into a single struct. While doing so it would also be good to remove the duplication of state currently present in `ServerHandshaker`: it keeps state both a state enum and a transcript, duplicate state can lead to unexpected behavior down the road, enabling security vulnerabilities.

Reply:
I choose not do this refactoring for this PR, as it necessitates significant changes to both the abstract attestation logic, and the grpc streaming implementation. However it is worth considering doing this refactoring after we remove the grpc streaming implementation.

Reply:
I agree that this is not optimal, but I think it is OK for an initial implementation. There is probably a need for a joint refactoring of the streaming and unary versions that is too big for this PR. Perhaps add an issue and a TODO.

Reply:
>  but I think it is OK for an initial implementation
agreed, hence why I implemented it this way. Will file an issue where we can aggregate refactoring needs